fix(linter/no-extraneous-class): improve docs, reporting and code refactor

DonIsaac opened the pull request:
## What This PR Does
Deviates from `@typescript-eslint/no-extraneous-class` by allowing empty, decorated classes by default.

This provides a better experience for NestJS user using Oxlint without a config file. In NestJS, [`Module`s are often empty but use decorators to declaratively state their contents](https://docs.nestjs.com/modules#feature-modules). Other similar frameworks (TSeD, maybe Angular) will also benefit.

### Other Changes
- Fix span on empty class diagnostics to not cover decorators.
- Provide help messages on all diagnostics
- Add missing links and rich text stuff to docs

DonIsaac commented:
* **#10797** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10797?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/10797?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=10797).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sysix commented:
I do not like that we are changing the default values. 
Users expect the migrated rules to work like the native rules.
`oxlint-migrate` needs to handle this case too, if we want to change this :/

camc314 commented on crates/oxc_linter/src/rules/typescript/no_extraneous_class.rs:
```diff
@@ -5,40 +5,60 @@ use oxc_ast::{
+            allow_constructor_only: false,
+            allow_empty: false,
+            allow_static_only: false,
+            allow_with_decorator: true,
```
:+1: to @Sysix's point, the defaults are all false here:

https://github.com/typescript-eslint/typescript-eslint/blob/a3a157c82a94fb61f4abaf8c4f42d51a7a78fb5a/packages/eslint-plugin/src/rules/no-extraneous-class.ts#L61-L64

DonIsaac committed:
fix(linter/no-extraneous-class): improve docs, reporting and code refactor (#10797)

## What This PR Does
Deviates from `@typescript-eslint/no-extraneous-class` by allowing empty, decorated classes by default.

This provides a better experience for NestJS user using Oxlint without a config file. In NestJS, [`Module`s are often empty but use decorators to declaratively state their contents](https://docs.nestjs.com/modules#feature-modules). Other similar frameworks (TSeD, maybe Angular) will also benefit.

### Other Changes
- Fix span on empty class diagnostics to not cover decorators.
- Provide help messages on all diagnostics
- Add missing links and rich text stuff to docs
